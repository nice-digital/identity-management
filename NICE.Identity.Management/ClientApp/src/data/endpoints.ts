const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
