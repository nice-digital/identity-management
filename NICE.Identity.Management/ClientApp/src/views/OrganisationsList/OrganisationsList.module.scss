.organisationsListSummary {
  font-size: 1.25rem;
  margin: 0;
}

.flex {
  display: flex;

  &SpaceBetween {
    justify-content: space-between;
  }
}
