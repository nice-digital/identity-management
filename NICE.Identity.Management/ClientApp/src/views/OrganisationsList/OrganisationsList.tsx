import React, { FC, useEffect, useState } from "react";
import { Link, Redirect, useLocation } from "react-router-dom";
import { useListFetch, isError } from "src/helpers/useListFetch";
import { useListInfo } from "src/helpers/useListInfo";
import { Endpoints } from "src/data/endpoints";
import { OrganisationType } from "src/models/types";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Button } from "@nice-digital/nds-button";
import { EnhancedPagination } from "@nice-digital/nds-enhanced-pagination";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { PageHeader } from "@nice-digital/nds-page-header";
import { ErrorMessage } from "src/components/ErrorMessage/ErrorMessage";
import { FilterSearch } from "src/components/FilterSearch/FilterSearch";
import { ResultsList } from "src/components/ResultsList/ResultsList";
import { SortOptions } from "src/components/SortOptions/SortOptions";
import { ItemsPerPage } from "src/components/ItemsPerPage/ItemsPerPage";
import { ResultOrganisation } from "src/components/ResultOrganisation/ResultOrganisation";
import { PaginationText } from "src/components/PaginationText/PaginationText";
import styles from "./OrganisationsList.module.scss";

export const OrganisationsList: FC = () => {
	const [organisations, setOrganisations] =
		useState<Array<OrganisationType>>(Object);
	const [organisationsCount, setOrganisationsCount] = useState(0);
	const [isLoading, setIsLoading] = useState(false);
	const [error, setError] = useState<Error | null>(null);
	const { pageNumber, totalPages, outOfRange, searchQuery } =
		useListInfo(organisationsCount);
	const listFetch = useListFetch<OrganisationType>();
	const { search: querystring } = useLocation();
	const querystringObject = new URLSearchParams(querystring);

	useEffect(() => {
		setIsLoading(true);
		(async () => {
			const listData = await listFetch(
				`${Endpoints.organisationsList}?q=${searchQuery || ""}`,
			);

			if (isError(listData)) {
				setError(listData.error);
			} else {
				setOrganisations(listData.items);
				setOrganisationsCount(listData.totalCount);
			}
			setIsLoading(false);
		})();
	}, [searchQuery, listFetch]);

	if (outOfRange) {
		querystringObject.set("page", "1");
		return <Redirect to={`/organisations?${querystringObject}`} />;
	}

	return (
		<>
			<Breadcrumbs>
				<Breadcrumb to="/overview" elementType={Link}>
					Administration
				</Breadcrumb>
				<Breadcrumb>Organisations</Breadcrumb>
			</Breadcrumbs>
			
			{!error ? (
				<>
					<PageHeader heading="Organisations" className="page-header mb--d" />
					<Grid>
						<GridItem cols={12}>
							<Button
								to={`/organisations/add`}
								variant="cta"
								elementType={Link}
								className="mb--e"
								data-qa-sel="add-organisation"
							>
								Add organisation
							</Button>
						</GridItem>
						<GridItem cols={12} md={3}>
							<FilterSearch label={"Filter by organisation name"} />
						</GridItem>
						<GridItem cols={12} md={9} aria-busy={true}>
							{isLoading ? (
								<p>Loading...</p>
							) : organisations.length ? (
								<>
									<div className="text-right">
										<SortOptions />
									</div>
									<PaginationText
										dataLength={organisationsCount}
										labelSingular="organisation"
										labelPlural="organisations"
									/>
									<ResultsList
										data={organisations}
										elementType={ResultOrganisation}
										qaSelExtract="organisations"
									/>
									<div className={`${styles.flex} ${styles.flexAlignCenter}`}>
										<ItemsPerPage />
										<EnhancedPagination
											currentPage={pageNumber}
											elementType={Link}
											method="to"
											mapPageNumberToHref={(pageNumber) => {
												const querystringObject = new URLSearchParams(
													querystring,
												);
												querystringObject.set("page", pageNumber.toString());
												return `?${querystringObject}`;
											}}
											totalPages={totalPages || 1}
											data-qa-sel="pagination-section"
										/>
									</div>
								</>
							) : searchQuery ? (
								<p>No results found for &quot;{searchQuery}&quot;</p>
							) : (
								<p>No results found</p>
							)}
						</GridItem>
					</Grid>
				</>
			) : (
				<>
					<PageHeader heading="Error" />
					<ErrorMessage error={error}></ErrorMessage>
				</>
			)}
		</>
	);
};
