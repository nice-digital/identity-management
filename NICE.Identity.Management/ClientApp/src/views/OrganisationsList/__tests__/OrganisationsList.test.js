import React from "react";
import { act } from "react-dom/test-utils";
import { render, waitFor, waitForElementToBeRemoved, fireEvent, screen, within } from "@testing-library/react";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { MemoryRouter } from "react-router";
import { OrganisationsList } from "../OrganisationsList";
import { Endpoints } from "src/data/endpoints";
import organisations from "./organisations.json";
import organisations2 from "./organisations2.json";

const consoleErrorReset = console.error;

beforeEach(() => {
	console.error = consoleErrorReset;
});

const server = setupServer(
  rest.get(Endpoints.organisationsList, (req, res, ctx) => {
    return res(ctx.json(organisations));
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("should show loading message before data has been loaded", () => {
	render(<OrganisationsList />, {wrapper: MemoryRouter});
	const loadingMessage = screen.getByText("Loading...", { selector: "p" });
	expect(loadingMessage).toBeInTheDocument();
});

test("should match the snapshot after data has been loaded", async () => {
	const {container} = render(<OrganisationsList />, {wrapper: MemoryRouter});	
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	expect(container).toMatchSnapshot();
});

test("should show error message when fetch organisations returns 401 error", async () => {
	console.error = jest.fn();	
	server.use(
		rest.get(Endpoints.organisationsList, (req, res, ctx) => {
		return res.once(
				ctx.status(401),
				ctx.json({})
			);
		})
	);
	const {container} = render(<OrganisationsList />, {wrapper: MemoryRouter});
	await waitFor(() => screen.getByRole("heading", { name: "Error"}));
	expect(container).toMatchSnapshot();	
});

test("should show error message when fetch organisations returns 500 error", async () => {
	console.error = jest.fn();	
	server.use(
		rest.get(Endpoints.organisationsList, (req, res, ctx) => {
			return res.once(
				ctx.status(500),
				ctx.json({})
			);
		})
	);
	const {container} = render(<OrganisationsList />, {wrapper: MemoryRouter});
	await waitFor(() => screen.getByRole("heading", { name: "Error"}));
	expect(container).toMatchSnapshot();	
});

test("should show no results message when initial fetch returns an empty array", async () => {
	server.use(
		rest.get(Endpoints.organisationsList, (req, res, ctx) => {
			return res.once(
				ctx.json([])
			);
		})
	);
	render(<OrganisationsList />, {wrapper: MemoryRouter});
	const noResultsMessage = await screen.findByText("No results found", { selector: "p" });
	expect(noResultsMessage).toBeInTheDocument();
});

test("should show no results message when search fetch returns an empty array", async () => {
	jest.useFakeTimers();
	const dummyText = "SomeText";
	server.use(
		rest.get(Endpoints.organisationsList, (req, res, ctx) => {
			const search = req.url.searchParams.get('q')
			
			if (search === dummyText) {
				return res.once(
					ctx.json([])
				);
			} else {
				return res(ctx.json(organisations));				
			}
		})
	);
	render(<OrganisationsList />, {wrapper: MemoryRouter});
	const searchInput = screen.getByLabelText("Filter by organisation name");
	searchInput.focus();
	await act(async () => {
		fireEvent.change(searchInput, {target: {value: dummyText}});
		jest.advanceTimersByTime(1000);
	});
	const noResultsFoundMessage = await screen.findByText(`No results found for "${dummyText}"`);
	expect(noResultsFoundMessage).toBeInTheDocument();
	jest.useRealTimers();
});

test("should show 25 (default page amount) or less results by default when paginated", async () => {
	server.use(
		rest.get(Endpoints.organisationsList, (req, res, ctx) => {
			return res.once(
				ctx.json(organisations2)
			);
		})
	);
	render(<OrganisationsList />, {wrapper: MemoryRouter});
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const organisationsList = screen.getByRole("list", { name: "organisations"});
	const { getAllByRole } = within(organisationsList);
	const organisationsListItems = getAllByRole("listitem");
	expect(organisationsListItems.length).toEqual(25);
});

test("should show all results when querystring amount is set as 'all'", async () => {
	server.use(
		rest.get(Endpoints.organisationsList, (req, res, ctx) => {
			return res.once(
				ctx.json(organisations2)
			);
		})
	);
	render(<MemoryRouter initialEntries={['/organisations?amount=all']}><OrganisationsList /></MemoryRouter>);
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const organisationsList = screen.getByRole("list", { name: "organisations"});
	const { getAllByRole } = within(organisationsList);
	const organisationsListItems = getAllByRole("listitem");
	expect(organisationsListItems.length).toEqual(organisations2.length);
});

test("should show page 2 when querystring page is set to 2", async () => {
	server.use(
		rest.get(Endpoints.organisationsList, (req, res, ctx) => {
			return res.once(
				ctx.json(organisations2)
			);
		})
	);
	render(<MemoryRouter initialEntries={['/organisations?page=2']}><OrganisationsList /></MemoryRouter>);
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const organisationsListSummary = await screen.findByText("Showing 26 to 32 of 32 organisations", { selector: "h2" });
	const organisationsList = screen.getByRole("list", { name: "organisations"});
	const { getAllByRole } = within(organisationsList);
	const organisationsListItems = getAllByRole("listitem");
	expect(organisationsListSummary).toBeInTheDocument();
	expect(organisationsListItems.length).toEqual(7);
});

test("should sort results by alphabetical descending when querystring sort is set to alpha-desc", async () => {
	server.use(
		rest.get(Endpoints.organisationsList, (req, res, ctx) => {
			return res.once(
				ctx.json(organisations2)
			);
		})
	);
	render(<MemoryRouter initialEntries={['/organisations?sort=alpha-desc']}><OrganisationsList /></MemoryRouter>);
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const organisationsList = screen.getByRole("list", { name: "organisations"});
	const { getAllByRole } = within(organisationsList);
	const organisationsListItems = getAllByRole("listitem");
	const { getByText } = within(organisationsListItems[0]);
	const firstOrgListItem = getByText("Zulu", { selector: "a" });
	expect(firstOrgListItem).toBeInTheDocument();
});