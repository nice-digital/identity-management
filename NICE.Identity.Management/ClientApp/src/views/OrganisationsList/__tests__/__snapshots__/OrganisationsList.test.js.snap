// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganisationsList should match the snapshot after data has been loaded 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Organisations
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Organisations\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header mb--d"
  >
    <h1
      className="page-header__heading"
    >
      Organisations
    </h1>
  </div>,
  <div
    className="grid"
  >
    <div
      data-g="12"
    >
      <a
        className="btn btn--cta mb--e"
        data-qa-sel="add-organisation"
        href="/organisations/add"
        onClick={[Function]}
      >
        Add organisation
      </a>
    </div>
    <div
      data-g="12 md:3"
    >
      <div
        className="panel"
      >
        <div
          className="input"
        >
          <label
            className="input__label"
            htmlFor="filter-search"
          >
            Filter by organisation name
          </label>
          <input
            autoComplete="off"
            className="input__input"
            data-qa-sel="filter-search-input"
            id="filter-search"
            name="filter-search"
            onChange={[Function]}
            type="search"
            unique="filterSearch"
          />
        </div>
      </div>
    </div>
    <div
      aria-busy={true}
      data-g="12 md:9"
    >
      <div
        className="text-right"
      >
        <ul
          className="list list--piped"
        >
          <li>
            <span
              className="pv--c pr--c"
            >
              A - Z
            </span>
          </li>
          <li>
            <a
              data-qa-sel="sort-alpha-desc"
              data-sort="alpha-desc"
              onClick={[Function]}
            >
              Z - A
            </a>
          </li>
          <li>
            <a
              data-qa-sel="sort-date-asc"
              data-sort="date-asc"
              onClick={[Function]}
            >
              Date ascending
            </a>
          </li>
          <li>
            <a
              data-qa-sel="sort-date-desc"
              data-sort="date-desc"
              onClick={[Function]}
            >
              Date descending
            </a>
          </li>
        </ul>
      </div>
      <h2
        className="organisationsListSummary"
        data-qa-sel="organisations-returned"
      >
        Showing 4 organisations
      </h2>
      <ul
        className="list--unstyled"
        data-qa-sel="list-of-organisations"
      >
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  <a
                    href="/organisations/1001"
                    onClick={[Function]}
                  >
                    NICE
                  </a>
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Date created
                  </dt>
                  <dd>
                    01-12-2021
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  <a
                    href="/organisations/1003"
                    onClick={[Function]}
                  >
                    Org 1
                  </a>
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Date created
                  </dt>
                  <dd>
                    11-01-2022
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  <a
                    href="/organisations/1004"
                    onClick={[Function]}
                  >
                    Org 2
                  </a>
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Date created
                  </dt>
                  <dd>
                    12-01-2022
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  <a
                    href="/organisations/1005"
                    onClick={[Function]}
                  >
                    Org 3
                  </a>
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Date created
                  </dt>
                  <dd>
                    13-01-2022
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
      </ul>
      <div
        className="flex flexAlignCenter"
      >
        <div
          className="mr--e"
          data-qa-sel="number-of-results-on-page"
        >
          <label
            className="bold mr--c"
            htmlFor="itemsPerPage"
          >
            Items per page
          </label>
          <select
            data-qa-sel="result-on-the-page-index"
            id="itemsPerPage"
            name="itemsPerPage"
            onChange={[Function]}
            value="all"
          >
            <option
              value="25"
            >
              25
            </option>
            <option
              value="50"
            >
              50
            </option>
            <option
              value="all"
            >
              All
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`OrganisationsList should show error message when fetch organisations returns 401 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Organisations
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Organisations\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header mb--d"
  >
    <h1
      className="page-header__heading"
    >
      Organisations
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;

exports[`OrganisationsList should show error message when fetch organisations returns 500 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Organisations
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Organisations\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header mb--d"
  >
    <h1
      className="page-header__heading"
    >
      Organisations
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;
